### DISCUSSION TITLE
Automatic static optimization should work with getStaticProps w/out getStaticPaths

### DISCUSSION THREAD

**Discussion by nicholaschiang** [vercel/next.js#22523] [Ideas]
**Describe the feature you'd like to request**

Automatic static optimization should work with `getStaticProps` w/out `getStaticPaths` when using dynamic SSG pages.

Right now, I have to add an empty `getStaticPaths` to every one of my dynamic pages (e.g. `pages/[org]/users/[id]/index.tsx`) that uses `getStaticProps`:

```typescript
export const getStaticPaths: GetStaticPaths<UserDisplayPageQuery> = async () => {
  return { paths: [], fallback: true };
};
```

This works fine, but is inconvenient and adds boilerplate code to each of my pages.

**Describe the solution you'd like**

Instead, Next.js should extend it's automatic static optimization capabilities to include dynamic pages with `getStaticProps` by simply using the existing `fallback: true` behavior. I don't know why this doesn't already exist. Seems easy enough.

**Describe alternatives you've considered**

Just adding that boilerplate `getStaticPaths` to every page:

```typescript
export const getStaticPaths: GetStaticPaths<UserDisplayPageQuery> = async () => {
  return { paths: [], fallback: true };
};
```

**Comment by TommySorensen** [comment]
Why not use the [optional catch all route](https://nextjs.org/docs/routing/dynamic-routes#optional-catch-all-routes) to catch multiple urls in the id file like this `pages/[org]/users/[[...id]].tsx`?
The point of using getStaticPaths is to fetch stuff non user specific content at build time or when page is requested. If it's not generated it will be generated in fallback mode until the page is generated. If you dont know the url for the `[org]` and `[id]` before user visit the url, then you dont need the "fallback" mode because there is nothing to prebuild for the NextJS.

When looking at https://github.com/tutorbookapp/tutorbook/blob/develop/pages/%5Borg%5D/users/%5Bid%5D/index.tsx#L128 it seems like you fetch user specific data in the `getStaticProps` function. As you [comment explain](https://github.com/tutorbookapp/tutorbook/blob/develop/pages/%5Borg%5D/users/%5Bid%5D/index.tsx#L136) there will be a "data change flashes". I will strongly recommend you not to do that since you are basically giving one users content to the next user because its static generated. So potentially you could run into some security issues here. Instead you should use `getServerSideProps` for that since you need the data at request time, and not for the last user's visit. Any user specific data could and should be fetched after hydration with useSwr.

This section explains perfect what i mean https://nextjs.org/docs/basic-features/data-fetching#fetching-data-on-the-client-side :
> If your page contains frequently updating data, and you don’t need to pre-render the data, you can fetch the data on the  client side. An example of this is user-specific data. Here’s how it works:
First, immediately show the page without data. Parts of the page can be pre-rendered using Static Generation. You can show loading states for missing data.
Then, fetch the data on the client side and display it when ready.
This approach works well for user dashboard pages, for example. Because a dashboard is a private, user-specific page, SEO is not relevant and the page doesn’t need to be pre-rendered. The data is frequently updated, which requires request-time data fetching.

**Comment by nicholaschiang** [comment]
@TommySorensen your concerns make sense, but I've already addressed all of them.

>  The point of using getStaticPaths is to fetch stuff non user specific content at build time or when page is requested. If it's not generated it will be generated in fallback mode until the page is generated. If you dont know the url for the [org] and [id] before user visit the url, then you dont need the "fallback" mode because there is nothing to prebuild for the NextJS.

I *was* using automatic static optimization (as described in [the client-side data fetching guide](https://nextjs.org/docs/basic-features/data-fetching#fetching-data-on-the-client-side)) but I want to show a list of organizations in the footer of each page.

Thus, I need to fetch that (public) organization data to render the footer links for each page. I'll fetch the data during build-time because it doesn't change often and I want it there immediately for SEO purposes.

> When looking at https://github.com/tutorbookapp/tutorbook/blob/develop/pages/%5Borg%5D/users/%5Bid%5D/index.tsx#L128 it seems like you fetch user specific data in the getStaticProps function. As you comment explain there will be a "data change flashes". I will strongly recommend you not to do that since you are basically giving one users content to the next user because its static generated. So potentially you could run into some security issues here. Instead you should use getServerSideProps for that since you need the data at request time, and not for the last user's visit. Any user specific data could and should be fetched after hydration with useSwr.

The data that is fetched in that `getStaticProps` function is public. The data change flashes occur if you're an admin and have access to the un-truncated data (which is fetched client-side using SWR if you're authenticated).

For user-specific pages (e.g. `pages/[org]/overview` or `pages/[org]/calendar`), I fetch sensitive data client-side using SWR; only authenticated users are able to see their data. But, I want to use `getStaticProps` on those pages to--as mentioned earlier--include a list of organizations in the page footer.

Does that make sense @TommySorensen? Thanks for responding BTW. Always nice to get a second opinion!

**Comment by nicholaschiang** [comment]
Related to #14200 because I want to be able to fetch data at build-time using `getStaticProps` that has nothing to do with the dynamic page parameters (i.e. page should behave just like an automatically statically optimized one).
